### PR TITLE
ESPnet-SPK: add ASVspoof19 SASV recipe

### PR DESCRIPTION
**Pull request by Alexgichamba**
## What?

Add a spk1 recipe on the ASVspoof19 Logical Access ASV subset.

## Why?

Provide diverse recipes for ESPnet-SPK, adding an SASV example.

## See also

[ASVspoof19](https://www.asvspoof.org/index2019.html)

### CODE REVIEW

**Comment by sw005320** [comment]
@Jungjee, can you review this?

File: egs2/TEMPLATE/spk1/db.sh [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Set the path of your corpus
 # "downloads" means the corpus can be downloaded by the recipe automatically
 
+ASVSpoof_LA=downloads
```
Comment (Jungjee) [inline egs2/TEMPLATE/spk1/db.sh:4]: Could you change the name to ASVSPOOF2019_LA or ASVSpoof2019_LA?

File: egs2/asvspoof/spk1/local/asv_data_prep.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# asv_data_prep.py
```
Comment (Jungjee) [inline egs2/asvspoof/spk1/local/asv_data_prep.py:1]: could you replace this line with a brief explanation what this script does?

File: egs2/asvspoof/spk1/local/asv_data_prep.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        spk2utt[speaker_id].append(utt_id)
+        utt2spk.append([utt_id, speaker_id])
+        wav_list.append([utt_id, full_path])
+
+    with open(os.path.join(dst, "spk2utt"), "w") as f_spk2utt, open(
+        os.path.join(dst, "utt2spk"), "w"
+    ) as f_utt2spk, open(os.path.join(dst, "wav.scp"), "w") as f_wav:
+
+        for spk in spk2utt:
+            f_spk2utt.write(f"{spk} {' '.join(spk2utt[spk])}\n")
```
Comment (Jungjee) [inline egs2/asvspoof/spk1/local/asv_data_prep.py:35]: note that there's an utt2spk to spk2utt script within ESPnet.

File: egs2/asvspoof/spk1/local/cm_data_prep.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
```
Comment (Jungjee) [inline egs2/asvspoof/spk1/local/cm_data_prep.py:1]: [question] Where do we need this? 
Do we need the train and the validation sets as well?

File: egs2/asvspoof/spk1/local/data.sh [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        # Make new protocol file
+        python local/convert_protocol.py --in_file "${data_dir_prefix}/LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.eval.gi.trl.txt" --out_file "${data_dir_prefix}/LA_asv_eval/protocol.txt"
+
+    else
+       log "LA_asv_eval exists. Skipping protocol modification"
+    fi
+    log "Stage 2, DONE."
+fi
+
+log "Successfully finished. [elapsed=${SECONDS}s]"
```
Comment (Jungjee) [inline egs2/asvspoof/spk1/local/data.sh:91]: remove?

File: egs2/asvspoof/spk1/run.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+set -e
+set -u
+set -o pipefail
+
+
+spk_config=conf/train_rawnet3.yaml
+
+train_set="train"
+valid_set="test"
+cohort_set="test"
```
Comment (Jungjee) [inline egs2/asvspoof/spk1/run.sh:11]: Please note that the cohort set cannot be equal to the test set. 
Could you use the development set since ASVspoof2019 LA explicitly has this partition?

**Review by Jungjee — COMMENTED** [review]
Thank you for the work Alex!
I left a few comments, but in general it looks great!

**Comment by Alexgichamba** [comment]
> Thank you for the work, Alex! I left a few comments, but in general, it looks great!
> 
> Meanwhile, please make additional required updates to make all CI tests pass.

Hi @Jungjee ,
Thanks for the review. I will address the comments.

File: egs2/asvspoof/spk1/local/cm_data_prep.py (hunk 4 above)
Comment (Alexgichamba) [reply egs2/asvspoof/spk1/local/cm_data_prep.py:1]: No. I did not use this. I thought perhaps someone using the recipe might want to do some in-domain finetuning. I could take it out if it is not needed.

File: egs2/asvspoof/spk1/run.sh (hunk 6 above)
Comment (Alexgichamba) [reply egs2/asvspoof/spk1/run.sh]: I see. Thanks for making me aware of this. I have added support for the dev partition, and I've assigned it to be the valid set and cohort set

**Comment by Alexgichamba** [comment]
Hi @Jungjee,
I have addressed the comments, except for two of them.

1. Should I remove the CM data prep? I had figured it might be useful later for in-domain finetuning
2. For the utt2spk/spk2utt, should I change it and use the script instead?
3. The CI checks that I am failing seem to be from install_espnet or test_python/test/espnet2/gan_tts. I am not quite sure how to make these pass

**Comment by sw005320** [comment]
> 3. The CI checks that I am failing seem to be from install_espnet or test_python/test/espnet2/gan_tts. I am not quite sure how to make these pass

I reran the CI

**Comment by Alexgichamba** [comment]
> > 3. The CI checks that I am failing seem to be from install_espnet or test_python/test/espnet2/gan_tts. I am not quite sure how to make these pass
> 
> I reran the CI

Thank you. Seems to be okay now

**Comment by sw005320** [comment]
> Thanks @Alexgichamba for the updates and your effort!!
> 
> > 1. Should I remove the CM data prep? I had figured it might be useful later for in-domain fine-tuning
> 
> I see. For me, it's okay to leave, but not so sure how @sw005320 would think about this.
> 

We can leave it, then.

**Comment by sw005320** [comment]
Thanks, @Alexgichamba
If you have some follow-up activities, please make a PR.

**Comment by Alexgichamba** [comment]
> Thanks, @Alexgichamba If you have some follow-up activities, please make a PR.

Thanks @sw005320 and @Jungjee. I appreciate the support!
